sam/bamRead.go

```diff
@@ -207,17 +219,31 @@ func DecodeBam(r *BamReader, s *Sam) (binId uint32, err error) {
 		qual[i] += 33 // ascii offset for printable characters
 	}
 
-	// s.Qual = unsafeByteToString(qual) // unsafe version
+	// ******
+	// Unsafe String Conversion
+	/*
+		s.Qual = unsafeByteToString(s.Qual, qual) // unsafe version
+	*/
+	// ******
+
+	// ******
+	// Safe String Conversion
 	s.Qual = string(qual) // TODO this is 1 alloc per read, should change to []byte and remove unsafe ref above
+	// ******
 
 	// The sam.Extra field is not parsed here as it would require parsing tags to their value, then
 	// casting that value to a string. That would be excessively wasteful, so instead the bytes are
 	// saved in the unparsedExtra field of sam, such that a user may call a function to parse these
 	// bytes if and only if they need access to the tag fields. This should also make bam reading a
 	// fair bit faster.
-	s.unparsedExtra = r.next(blkSize - (staticBamAlnSize +
+	ex := r.next(blkSize - (staticBamAlnSize +
 		lenReadName + (4 * numCigarOps) + (((lenSeq) + 1) / 2) + lenSeq)) // to get remaining bytes in alignment
-
+	if cap(s.unparsedExtra) < len(ex) {
+		s.unparsedExtra = make([]byte, len(ex))
+		copy(s.unparsedExtra[:len(ex)], ex)
+	}
+	s.unparsedExtra = s.unparsedExtra[:len(ex)]
```

Comment:
I think you want to wrap this line in an "else" because you don't need to do it if you went into the above "if"